sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs

```diff
@@ -873,5 +1063,277 @@ public virtual Task FlushInternalAsync(CancellationToken cancellationToken = def
+                }
+                catch (TaskCanceledException)
+                {
+                    // This is expected; no action is needed.
+                }
+                catch (Exception ex)
+                {
+                    Logger.BufferedProducerBackgroundProcessingStopError(Identifier, EventHubName, ex.Message);
+                    (capturedExceptions ??= new List<Exception>()).Add(ex);
+                }
+
+                // Clean up all partition publishers.
+
+                foreach (var pair in _activePartitionPublishers)
+                {
+                    // Dispose for the partition publishers will log exceptions and
+                    // avoid surfacing them, as processing is stopping.
+
+                    try
+                    {
+                        _activePartitionPublishers.TryRemove(pair.Key, out _);
+                        pair.Value.Dispose();
+                    }
+                    catch (Exception ex)
+                    {
+                       Logger.BufferedProducerBackgroundProcessingStopError(Identifier, EventHubName, ex.Message);
+                       (capturedExceptions ??= new List<Exception>()).Add(ex);
+                    }
+                }
+
+                _producerManagementTask = null;
+            }
+            catch (Exception ex)
+            {
+                if (ex is OperationCanceledException opEx)
+                {
+                    throw new TaskCanceledException(opEx.Message, opEx);
+                }
+
+                Logger.BufferedProducerBackgroundProcessingStopError(Identifier, EventHubName, ex.Message);
+                (capturedExceptions ??= new List<Exception>()).Add(ex);
+            }
+            finally
+            {
+                Logger.BufferedProducerBackgroundProcessingStopComplete(Identifier, EventHubName);
+            }
+
+            // Surface any exceptions that were captured during cleanup.
+
+            if (capturedExceptions?.Count == 1)
+            {
+                ExceptionDispatchInfo.Capture(capturedExceptions[0]).Throw();
+            }
+            else if (capturedExceptions is not null)
+            {
+                throw new AggregateException(capturedExceptions);
+            };
+        }
+
+        /// <summary>
+        ///   Performs the actions needed to validate that the partition state information for the
+        ///   producer is current, updating the class members if needed.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <remarks>
+        ///   This method will read and mutate class-level state for partitions using the
+        ///   shared producer client instance.
+        /// </remarks>
+        ///
+        private async Task EnsurePartitionStateAsync(CancellationToken cancellationToken = default)
```

Comment:
Do call to this need to be synchronized or is there already some upstream guarantee that we won't race to set 
 `_partitions` and `_partitionHash`?

Reply:
The intention is for this to only be called when processing is bootstrapped for the first time and from the background task.  Even were it to be called outside of that, I don't think there would be a concern.   The partition data may be slightly stale and we may not see new partitions in the rare event that they were added, but the next tick of the background thread would update.